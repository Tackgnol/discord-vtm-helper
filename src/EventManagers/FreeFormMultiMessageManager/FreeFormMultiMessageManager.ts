import { isEmpty, isNil, intersection } from 'lodash';
import { MessageEmbed } from 'discord.js';
import { settings } from '../../config/settings';
import { MessageList, Reply, ReplyType } from '../../Models/AppModels';
import { InvalidInputError } from '../../Common/Errors';

export class FreeFormMultiMessageManager {
	messageUsers(reply: string, userList: string[], channelMembers: string[]): Reply {
		const users = intersection(userList, channelMembers);
		if (isEmpty(users)) {
			throw new InvalidInputError('No users to send to');
		}
		const messageLists: MessageList[] = [];
		users.forEach(u => {
			const foundUser = channelMembers.find(val => val === u);
			if (!isNil(foundUser)) {
				const richEmbed = new MessageEmbed()
					.setColor(settings.colors.richEmbeddedMain)
					.setTitle(settings.lines.userMessageHeader)
					.setDescription(reply)
					.setColor(settings.colors.richEmbeddedMain);
				messageLists.push({ message: richEmbed, recipient: u });
			}
		});
		return {
			type: ReplyType.Multi,
			value: messageLists,
		};
	}
}
