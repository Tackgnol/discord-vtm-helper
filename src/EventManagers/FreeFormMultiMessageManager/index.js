const { isEmpty, isNil } = require('lodash');

class FreeFormMultiMessageManager {
	constructor(message, channel) {
		this.message = message;
		this.channel = channel;
	}

	messageUsers(reply, userList) {
		const messageChanel = this.channel;
		if (messageChanel.type === 'text') {
			const channelMembers = messageChanel.members;
			if (!isEmpty(userList)) {
				userList.forEach(u => {
					const foundUser = channelMembers.find(val => val.user.username === u);
					if (!isNil(foundUser)) {
						foundUser.send(reply);
					}
				});
			}
		}
	}
}

module.exports = FreeFormMultiMessageManager;
