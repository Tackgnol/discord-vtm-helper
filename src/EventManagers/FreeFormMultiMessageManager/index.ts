import { isEmpty, isNil } from 'lodash';
import { Message, RichEmbed, TextChannel, User } from 'discord.js';
import { settings } from '../../config/settings';

export class FreeFormMultiMessageManager {
	constructor(private readonly channel: TextChannel) {
		this.channel = channel;
	}

	messageUsers(reply: Message, userList: User[]) {
		const messageChanel = this.channel;
		if (messageChanel.type === 'text') {
			const channelMembers = messageChanel.members;
			if (!isEmpty(userList)) {
				userList.forEach(u => {
					const foundUser = channelMembers.find(val => val.user.username === u.username);
					if (!isNil(foundUser)) {
						const richEmbed = new RichEmbed()
							.setColor(settings.colors.richEmbeddedMain)
							.setTitle(settings.lines.userMessageHeader)
							.setDescription(reply.content)
							.setColor(settings.colors.richEmbeddedMain);
						foundUser.send(richEmbed);
					}
				});
			}
		}
	}
}
