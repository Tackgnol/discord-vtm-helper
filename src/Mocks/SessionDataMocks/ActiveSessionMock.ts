import { IActiveSession } from '../../Models/AppModels';
import { existingGTQueryWithoutValue } from '../QueryMocks/globalTestQueries';

export const activeSessionMock: IActiveSession = {
	channelId: '0',
	gameId: '0',
	prevCommand: existingGTQueryWithoutValue,
};
