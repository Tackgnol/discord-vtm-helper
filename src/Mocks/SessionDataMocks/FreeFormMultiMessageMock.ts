export const freeFormMultiMessageMock = {
	reply: 'test',
	userList: ['1', '2', '3'],
	channelMembers: ['2', '3', '4'],
};

export const freeFormMultiMessageErrorMock = {
	reply: 'test',
	userList: ['1', '2', '3'],
	channelMembers: ['4', '5', '6'],
};
