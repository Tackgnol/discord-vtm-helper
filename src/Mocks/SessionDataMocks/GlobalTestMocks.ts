import { GlobalTest } from '../../Models/GameData';

export const globalTestWithoutShortCircuitMock: GlobalTest = {
	testMessage: 'test',
	shortCircuit: false,
	globaltestoptionSet: [
		{ minResult: 1, resultMessage: 'result for 1' },
		{ minResult: 5, resultMessage: 'result for 5' },
	],
	name: 'test',
	replyPrefix: 'testPrefix',
};
export const globalTestWithShortCircuitMock: GlobalTest = {
	testMessage: 'shortCircuitTest',
	shortCircuit: true,
	globaltestoptionSet: [
		{ minResult: 1, resultMessage: 'result for 1' },
		{ minResult: 5, resultMessage: 'result for 5' },
	],
	name: 'scTest',
	replyPrefix: 'testPrefix',
};
