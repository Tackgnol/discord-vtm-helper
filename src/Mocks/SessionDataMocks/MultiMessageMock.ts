import { IMultiPlayerMessage } from '../../Models/GameData';

export const multiMessageMock: IMultiPlayerMessage = {
	userList: ['0'],
	name: 'test',
	message: "you've got mail",
};
