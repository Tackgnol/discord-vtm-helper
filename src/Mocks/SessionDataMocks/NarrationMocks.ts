import { INarration } from '../../Models/GameData';

export const basicNarration: INarration = {
	name: 'test',
	image: 'urlToImage',
	narrationText: 'this is a narration text',
};
