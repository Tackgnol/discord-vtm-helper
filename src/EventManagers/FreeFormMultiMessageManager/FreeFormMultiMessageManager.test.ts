import { FreeFormMultiMessageManager } from './FreeFormMultiMessageManager';
import { freeFormMultiMessageErrorMock, freeFormMultiMessageMock } from '../../Mocks/SessionDataMocks/FreeFormMultiMessageMock';
import chai, { expect } from 'chai';
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot';

chai.use(jestSnapshotPlugin());

describe('EventManagers >> FreeFormMultiMessageManager >> FreeFormMultiMessageManager', () => {
	const manager = new FreeFormMultiMessageManager();

	it('returns a proper message list', () => {
		const { channelMembers, reply, userList } = freeFormMultiMessageMock;
		const result = manager.messageUsers(reply, userList, channelMembers);
		expect(result).toMatchSnapshot();
	});

	it('throws an error on empty user lists', () => {
		const { channelMembers, reply } = freeFormMultiMessageMock;
		expect(() => {
			manager.messageUsers(reply, [], channelMembers);
		}).to.throw;
	});

	it('throws an error on empty channel members', () => {
		const { reply, userList } = freeFormMultiMessageMock;
		expect(() => {
			manager.messageUsers(reply, userList, []);
		}).to.throw;
	});

	it('throws an error if the channel list and user list do not match', () => {
		const { channelMembers, reply, userList } = freeFormMultiMessageErrorMock;
		expect(() => {
			manager.messageUsers(reply, userList, channelMembers);
		}).to.throw;
	});
});
