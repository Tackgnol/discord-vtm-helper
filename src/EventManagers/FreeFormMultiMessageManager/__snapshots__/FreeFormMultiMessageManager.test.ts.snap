// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventManagers >> FreeFormMultiMessageManager >> FreeFormMultiMessageManager returns a proper message list 1`] = `
Object {
  "type": 2,
  "value": Array [
    Object {
      "message": Object {
        "author": null,
        "color": 9044739,
        "description": "test",
        "fields": Array [],
        "footer": null,
        "image": null,
        "thumbnail": null,
        "timestamp": null,
        "title": "Whispers in the dark... be it friend or foe or thyself?",
        "type": "rich",
        "url": null,
      },
      "recipient": "2",
    },
    Object {
      "message": Object {
        "author": null,
        "color": 9044739,
        "description": "test",
        "fields": Array [],
        "footer": null,
        "image": null,
        "thumbnail": null,
        "timestamp": null,
        "title": "Whispers in the dark... be it friend or foe or thyself?",
        "type": "rich",
        "url": null,
      },
      "recipient": "3",
    },
  ],
}
`;
